jax/interpreters/pxla.py

```diff
@@ -331,32 +387,42 @@ class ShardedDeviceArray(xla.DeviceArray):
   behavior of an ndarray so that it can be treated by user code as an ndarray;
   that is, it is only an optimization to reduce transfers.
 
-  The number of device buffers underlying a ShardedDeviceArray instance is equal
-  to the number of replicas of the computation that produced it. Each buffer
-  represents a shard of the original array, meaning a slice along its leading
-  axis. These component buffers reside on distinct devices, but need not
-  represent distinct logical shards. The correspondence can be computed with
-  the assign_shards_to_replicas function.
+  Attributes:
+    aval: A ShapedArray indicating the shape and dtype of this array.
+    sharding_spec: describes how this array is sharded across `device_buffers`.
+    device_buffers: the buffers containing the data for this array. Each buffer
+      is the same shape and on a different device. Buffers are in row-major
+      order, with replication treated as an extra innermost dimension. For
+      example, if the unsharded data is [1, 2], the replicated data should be
+      laid out here as [1, 1, 2, 2].
+    indices: the result of spec_to_indices(sharding_spec). Can optionally be
+      precomputed for efficiency. A list the same length as `device_buffers`
+      (equal to `prod(sharding_spec.shards_per_axis)`). Each index indicates what
+      portion of the full array is stored in the corresponding device buffer,
+      i.e. `array[indices[i]] == device_buffers[i].to_py()`.
   """
-  __slots__ = ["device_buffers", "axis_size"]
-  _collect = staticmethod(onp.stack)
-
-  def __init__(self, aval, device_buffers):
+  __slots__ = ["device_buffers", "sharding_spec", "indices"]
+
+  # TODO(skye): expose PyLocalBuffers in xla_client
+  def __init__(self,
+               aval: ShapedArray,
+               sharding_spec: ShardingSpec,
+               device_buffers: List[xb.xla_client._xla.PyLocalBuffer],
+               indices: List[Index] = None):
+    # TODO(skye): assert invariants. Keep performance in mind though.
+    if indices is None:
+      indices = spec_to_indices(aval.shape, sharding_spec)
     self.aval = aval
     self.device_buffers = device_buffers
-    self.axis_size = aval.shape[0]
+    self.sharding_spec = sharding_spec
+    self.indices = indices
     self._npy_value = None
     if not core.skip_checks:
       assert type(aval) is ShapedArray
 
-  def _ids(self):
-    num_bufs = len(self.device_buffers)
-    assignments = assign_shards_to_replicas(num_bufs, self.axis_size)
-    _, ids = onp.unique(assignments, return_index=True)
-    return ids
-
   def copy_to_host_async(self):
     if self._npy_value is None:
+      # TODO(skye): only transfer one replica?
```

Comment:
+1 🙂 
I think this would look pretty similar to the striding in `_value`?

Reply:
Yeah, it'd be easy. I just didn't wanna change the behavior in this PR, I'll do it as a follow-up on the off chance anyone is depending on the current behavior.